File: thesync-frontend/types/notification.ts
export type NotificationType =
  | "approved"
  | "reminder"
  | "reschedule"
  | "completed"
  | "info";

export interface Notification {
  id: string;
  type: NotificationType;
  title: string;
  message: string;
  /** ISO 8601 date string — used for relative time display and sorting */
  timestamp: string;
  read: boolean;
  /** ID of the related consultation for deep linking */
  relatedConsultationId?: string;
}

export interface NotificationResponse {
  id: string;
  user_id: string;
  schedule_id?: string | null;
  message: string;
  is_read: boolean;
  created_at: string;
}

export interface NotificationListResponse {
  items: NotificationResponse[];
  total: number;
  page: number;
  page_size: number;
  total_unread: number;
}

File: thesync-backend/model/notification.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import NonNegativeInt, field_validator

from model.base import DomainModel, NonEmptyText, normalize_optional_text


class Notification(DomainModel):
    """User-facing notification emitted by schedule events."""

    id: UUID
    user_id: UUID
    schedule_id: UUID | None = None
    message: NonEmptyText
    is_read: bool = False
    created_at: datetime

    @field_validator("message", mode="before")
    @classmethod
    def normalize_message(cls, value: str) -> str:
        normalized = normalize_optional_text(value)
        if normalized is None:
            raise ValueError("message must not be empty.")

        return normalized


class NotificationResponse(Notification):
    """Notification response shape returned by the API."""


class NotificationListResponse(DomainModel):
    """Paginated notification response including unread totals."""

    items: list[NotificationResponse]
    total: NonNegativeInt
    page: NonNegativeInt
    page_size: NonNegativeInt
    total_unread: NonNegativeInt = 0


class NotificationMarkAllReadResult(DomainModel):
    """Result of marking notifications as read in bulk."""

    updated_count: NonNegativeInt = 0
